### ISSUE TITLE
BIDS derived datasets file naming

### ISSUE THREAD

**Issue by mariehbourget** [ivadomed/ivadomed#860]
## Disclaimer
This is a discussion issue based on my understanding of the BIDS specification for derivatives and on some of ivadomed’s pipelines, it may not be entirely accurate, hence the `design_discussion` and `help wanted` labels. Don't hesitate to correct me. I also don't have a lot of time available to look further into this at the moment but I thought it was worth starting a discussion.

## Context

Some of our pipelines in ivadomed use data that is not the original “raw data” but data that has been pre-processed in some way (e.g. extracting the mid slice for vertebral labelling (#707), averaging and motion correction on spine generic (https://github.com/spine-generic/spine-generic/pull/249))

Other similar situations are arising with small microscopy dataset where we want to split/crop a raw image in several parts, some of which would go to the training set, and others to the validation set. Therefore, each of those image “parts” would be considered “pre-processed” and would need to have and name and structure BIDS-compliant.

## What we have done so far

For MRI data, we started to use the entity `rec-` (reconstruction) to identify such pre-processed data.
```
awesome-data/sub-bob/anat/sub-bob_T1w.nii.gz                     --> goes into the pre-processing pipeline
awesome-data-processed/sub-bob/anat/sub-bob_rec-XX_T1w.nii.gz    --> goes into training
```

But, we have encountered some issues with this. Ex: `rec-` is not an admissible entity for `dwi` (#824) and won't be indexed by `pybids` without an exception. We were also not planning on using it in the Microscopy BEP.

Moreover, the more I read about this, the more I think we may be using it wrong, i.e. the `rec-` entity is part of the “raw data” specification, therefore should be used on raw data only. The example provided in the spec [here](https://bids-specification.readthedocs.io/en/stable/99-appendices/09-entities.html#rec) is a motion correction reconstruction algorithm, which from my understanding is not apply “after” the acquisition, but “during” the acquisition of the "raw data".

## Possible solution

I read a little bit more in the BIDS spec and found that there is a concept of a “derived dataset” (see [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets) in the example 2). This is a “stand-alone” derived dataset independent from the source BIDS dataset. 

Then in the [BIDS Derivatives](https://bids-specification.readthedocs.io/en/stable/05-derivatives/01-introduction.html) part of the spec, they explain how to name files in a derived dataset. It seems that different key/value pairs can be add to the filename that are specific to derivatives.

For example, the key `desc-` (description) that is used to distinguish files that would otherwise have the same name or to prevent the file having the same name as the original (example [here](https://bids-specification.readthedocs.io/en/stable/05-derivatives/02-common-data-types.html#examples))

This seems to be exactly what we would need to replace the `rec-` entity and have a BIDS-compliant derived dataset.
```
awesome-data/sub-bob/anat/sub-bob_T1w.nii.gz                      --> goes into the pre-processing pipeline
awesome-data-processed/sub-bob/anat/sub-bob_desc-XX_T1w.nii.gz    --> goes into training
```

Other “derivatives-specific” entities from this part of the spec may also be useful for us.

## Potential issues

I can foresee two issues with moving on with this possible solution:
1. We would have to make sure that the derived dataset entities (ex: `desc-`) are indexed correctly by `pybids` in our ivadomed pipelines.
2. We may have to rename several files in already existing BIDS dataset, which may break compatibility with previous experiment.

## Additional note
There is also a `DatasetType` key in `dataset_description.json` that should be used to identified “raw” vs. “derivative” datasets [here](https://bids-specification.readthedocs.io/en/stable/03-modality-agnostic-files.html#dataset_descriptionjson).

**Comment by jcohenadad** [comment]
@mariehbourget thank you for raising this much needed discussion and for describing the context and possible options in a very clear manner. I totally agree with that plan. So, IIUC, it would _also_ be possible to have a "desc" file inside an existing BIDS dataset. Example:

~~~
awesome-data/sub-bob/anat/sub-bob_desc-upsampled_T1w.nii.gz
~~~

I can think of a few situations where this would be useful.

Regarding the issues you raised, they seem manageable.

**Comment by mariehbourget** [comment]
> So, IIUC, it would _also_ be possible to have a "desc" file inside an existing BIDS dataset.

No, that's the twist. Some entities like `desc-`, `res-` (which we use in the Microscopy BEP) and a few others are only admissible for "derived datasets". See `desc-` [here](https://bids-specification.readthedocs.io/en/stable/99-appendices/09-entities.html#desc).

I think it would be a good idea to move forward in that direction but we have to make sure to understand correctly how the "derived dataset" concept is supposed to be used. It would help to have a second pair of eyes on this to review the fine details.

**Comment by jcohenadad** [comment]
So, quickly looking at the [BIDS website](derivative) "desc-" is applicable to "derivative data", which to my understanding is not _another_ BIDS dataset (as [exemplified](https://github.com/ivadomed/ivadomed/issues/860#issue-943568439) with "awesome-data-processed"), but in the _same_ BIDS dataset under the folder "derivatives/", which is supposed to include processed data (eg with FSL, SCT, etc.) and manual labels (eg: [here](https://github.com/spine-generic/data-multi-subject/tree/master/derivatives/labels)).

**Comment by mariehbourget** [comment]
From my understanding, the derivatives can be either:
1. in the derivatives folder of the `awesome_data` dataset
2. in a derived `awesome-data-processed` dataset as main subjects-folders (not in derivatives)

Both cases are explained [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets). For our use, I think case 2 is more suited.

**Example case 1:**
```
awesome-data/sub-bob/anat/sub-bob_T1w.nii.gz --> "raw data" not preprocessed
awesome-data/derivatives/awesome-software/sub-bob/anat/sub-bob_desc-XX_T1w.nii.gz --> pre-processed data with awesome software
awesome-data/derivatives/awesome-software/derivatives/labels/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz --> segmentation label of pre-processed data
```
- I don't know if the third line (derivatives in derivatives) is allowed in BIDS
- In ivadomed, the `bids_path` would then be `awesome-data/derivatives/awesome-software` and not `awesome-data`, this would have to be tested.

**Example case 2:**
```
awesome-data/sub-bob/anat/sub-bob_T1w.nii.gz --> "raw data" not preprocessed
awesome-data-prepocess/sub-bob/anat/sub-bob_desc-XX_T1w.nii.gz --> pre-processed data as main subjects-folder in derived dataset
awesome-data-prepocess/derivatives/labels/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz --> segmentation label of pre-processed data
```
- In ivadomed, the `bids_path` would then be `awesome-data-prepocess`, and the file indexation would have to be tested with the new entity `desc-`.

**Comment by jcohenadad** [comment]
You're absolutely right, thank you for the clarification

**Comment by hermancollin** [comment]
For confirmation, the `desc-` entity does get recognized by ivadomed when loading the dataset. Also, I didn't even have to specify that the dataset was a derived dataset (in _dataset_description.json_, "DatasetType" key was ommitted).

**Comment by mariehbourget** [comment]
> For confirmation, the desc- entity does get recognized by ivadomed when loading the dataset

Great, thank you @hermancollin for the test. Just to confirm, there is a column "description" in the `bids_dataframe.csv` with the right info?
We should do a similar test with MRI `anat` data as well to make sure everything is fine (microscopy files are forced indexed because they are not part of BIDs yet whereas `anat` data has to go through the BIDS-validator)

**Comment by hermancollin** [comment]
> Just to confirm, there is a column "description" in the `bids_dataframe.csv` with the right info?
Ah! Sorry forgot to look inside that file. Unfortunately there is not "description" column. The input files did load properly but apart from the path/filename fields, they are indistinguishable in the `bids_dataframe.csv` file...

**Comment by hermancollin** [comment]
Maybe the problem is that the `desc-` field is not in present in the bids config I used (default one):
https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config_bids.json

**Comment by mariehbourget** [comment]
> Maybe the problem is that the `desc-` field is missing from the default bids_config.json file

This "custom" config file is used to override the default config file from `pybids` (again, because microscopy is not BIDS yet)

In `pybids` [here](https://github.com/bids-standard/pybids/tree/master/bids/layout/config), are:
- the default config file (`bids.json`), this is the config used when we run ivadomed with MRI `anat` data (no `bids_config` parameter required in the ivadomed config file for that)
- the derivatives config file (`derivatives.json`), which includes the "derivatives-only" entities.

So the indexing relies entirely on `pybids` for the derivatives (I think!)
We should test:
- the same microscopy dataset with the addition of the "DatasetType" key in dataset_description.json
- a small MRI `anat` test dataset with `desc-` and with the "DatasetType" key

**Comment by hermancollin** [comment]
> So the indexing relies entirely on `pybids` for the derivatives (I think!)
> We should test:
> 
> * the same microscopy dataset with the addition of the "DatasetType" key in dataset_description.json

Moreover we might want to add the `_desc-` entity in the custom _config_bids.json_ file. It is described in the _derivative.json_ file from the `pybids` link you provided:
```json
{
    "name": "desc",
    "pattern": "desc-([a-zA-Z0-9]+)"
}
```

**Comment by mariehbourget** [comment]
> Moreover we might want to add the `_desc-` entity in the custom _config_bids.json_ file. It is described in the [_derivative.json_](https://github.com/bids-standard/pybids/blob/master/bids/layout/config/derivatives.json) file from the `pybids` link you provided:

Yeah maybe, but my understanding is that the `derivatives.json` should already do the job without having to specify it, but it's just a supposition at this point based on:
- `ivadomed` uses `pybids` to index
- `desc-` is not in the default `bids.json` from `pybids`
- `pybids` uses `bids.json` + `derivatives.json` + `BIDS-validator` for BIDS-compliant modalities (ex: MRI `anat`)
    - we have to confirm this with a test on `anat` data with `desc-` field.
- `pybids` uses our custom `config_bids.json` and skip the `BIDS-validator` for non BIDS-compliant modalities (ex: `microscopy`)
    -  the question is: what happens with `derivatives.json` in that case?

**Comment by hermancollin** [comment]
Just tried with the `"DatasetType"` field set to `"derivative"` and unfortunately the `desc` field is not added into the _bids_dataframe.csv_ file. Maybe this indicates that the `derivative.json` file isn't used as config file even when inputting a derived dataset into ivadomed but further investigation is needed.

**Comment by naga-karthik** [comment]
Thank you for the detailed description of this issue @mariehbourget! I totally agree with how you have presented the 2nd method of [storing the "derived" datasets](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets). I do have some questions in general, which are described below:
1. Am I correct in understanding that when we want to shift from using `rec-` to `desc-`, in addition to renaming the files, we also have to reorganize how the folders are stored? In particular, currently in `ivadomed`, the idea of the 2nd method of storing preprocessed (derived) datasets (see link above) is not implemented, rather, it is all inside a single BIDS compliant structure that has a `derivatives` folder that contains all the preprocessed data. _This_ has to be changed (and tested), right?
2. As for the current state of things, tests by @hermancollin confirm that the custom bids config file being used for microscopy data works with the `desc-` file name modification. However, it has to be tested for MRI `anat` data to be absolutely sure this works because of the third point in this [comment](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-880201659). Is this correct?

If point 1 makes sense, then one way to check whether this works with ivadomed is to create a small MRI dataset with separate `raw` and `derivatives` datasets like shown in the example code for the 2nd method [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets) and change all the occurrences of `rec-` to `desc-`. What do you think @mariehbourget ? Does it make sense? Lmk if you have any other suggestions. 

I also agree with your hunch that we do not have to tweak `pybids`'s  `bids.json` file by adding the `desc-` key because the `derivatives.json` includes it (as you already said). So, as far as I can understand, reorganizing the folder structure and renaming files should do it for the MRI test.

**Comment by mariehbourget** [comment]
Thank you @naga-karthik for looking into this.

> 1. Am I correct in understanding that when we want to shift from using `rec-` to `desc-`, in addition to renaming the files, we also have to reorganize how the folders are stored? In particular, currently in `ivadomed`, the idea of the 2nd method of storing preprocessed (derived) datasets (see link above) is not implemented, rather, it is all inside a single BIDS compliant structure that has a `derivatives` folder that contains all the preprocessed data. _This_ has to be changed (and tested), right?

Right now in `ivadomed`, we work with 2 dataset structures:
1. the "raw data" in `sub-XX` folders at the root of the dataset and "labels" in the `derivatives` folder.
2. the "preprocessed data" in `sub-XX` folders at the root of the dataset, with `rec-` in the filenames and "labels" in the `derivatives` folder.

We would like to do the exact same thing as the second one but with `desc-` in the filename instead of `rec-` (to be more BIDS compliant).

So for the 2nd method of storing derived dataset (the stand-alone independent version, example 2 in the link), we would not need to re-arrange the structure. We would have two things:
1. `awesome-data-preprocess/sub-bob/anat/sub-bob_desc-XX_T1w.nii.gz`  (same structure as what we have now but with `desc-` in the filename)
2. `awesome-data-preprocess/derivatives/labels/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz` : for our segmentation labels.

What is ambiguous here is that the `awesome-data-preprocess` dataset is already a "derivative" dataset (because it is not the "raw data" in BIDS terms), but we also have the labels in the `derivatives` folder (which are like a derivatives of derivatives).

> 2. As for the current state of things, tests by @hermancollin confirm that the custom bids config file being used for microscopy data works with the `desc-` file name modification. However, it has to be tested for MRI `anat` data to be absolutely sure this works because of the third point in this [comment](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-880201659). Is this correct?

You are right, currently the `desc-` entity is working for indexing microscopy files. However, the "description" column does not appear in the `bids_dataframe.csv`, which we think it should. We would like to test with MRI data because MRI is an official BIDS modality and will go through the real `pybids` indexation, without the custom `config_bids.json` file and without files being forced indexed by `ivadomed` like for microscopy at the moment.

> If point 1 makes sense, then one way to check whether this works with ivadomed is to create a small MRI dataset with separate `raw` and `derivatives` datasets like shown in the example code for the 2nd method [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets) and change all the occurrences of `rec-` to `desc-`. What do you think @mariehbourget ? Does it make sense? Lmk if you have any other suggestions.

Yes, we would need a small "preprocessed" MRI "stand-alone" dataset. Either use a dataset with the `rec-` entity and changing it to `desc-`, or we could take our `spinegeneric_example_dataset`, and add `desc-` to the name of a few files and derivatives.
We may also need to add the "DatasetType" key with `derivative` value in the `dataset_description.json` file of the dataset.
We would first need to check:
- Are the files indexed in the `bids_dataframe.csv`
- Is the column "description" (or maybe "desc") is present in the dataframe?

Let me know if you have questions or if I misunderstood something, thank you!

**Comment by naga-karthik** [comment]
> What is ambiguous here is that the awesome-data-preprocess dataset is already a "derivative" dataset (because it is not the "raw data" in BIDS terms), but we also have the labels in the derivatives folder 

Yeah exactly! Thanks for mentioning this. I was wondering the same thing about how the approach ivadomed has taken is a bit confusing in terms of using the folder name "derivatives". But, I understand now, we just have to rename `rec-` to `desc-` for the purpose of this issue.

> However, the "description" column does not appear in the `bids_dataframe.csv`, which we think it should.

Sorry but where can I find this csv file? I couldn't read about it on BIDS' documentation. Is it something exclusive to ivadomed?

> we could take our `spinegeneric_example_dataset`, and add desc- to the name of a few files and derivatives

I think we can go ahead with this approach as it seems simpler. 
At this point, I am only limited by my knowledge on the `bids_dataframe.csv` file. I can get back with some more suggestions/actions after I read about it some more.

**Comment by mariehbourget** [comment]
> Sorry but where can I find this csv file? I couldn't read about it on BIDS' documentation. Is it something exclusive to ivadomed?

Yes, this is in ivadomed, in the `path_output` folder, the loader writes a `bids_dataframe.csv` with all the files and metadata indexed and used from the `path_data` for a given training/testing, it came from `https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/main.py#L351` in the main.

**Comment by naga-karthik** [comment]
Based on some experiments and suggestions thanks to @mariehbourget, here is my latest finding:

* I created a copy of the `data_example_spinegeneric` available from [this](https://ivadomed.org/en/latest/tutorials/one_class_segmentation_2d_unet.html) tutorial and kept the names of three files as it is and renamed all the remaining files according to the following:
  *  `awesome-data-preprocess/sub-bob/anat/sub-bob_desc-XX_T1w.nii.gz` for the input training data
  *  `awesome-data-preprocess/derivatives/labels/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz` for the ground truth labels.
  * The `DatasetType`, specifically, `"DatasetType": "derivative"` was also added to both the `dataset_description.json` files in the respective root and `derivatives` folders.
  * The dataset's folder structure was kept unchanged. Only filenames were modified to include `desc-`. 

**Observation**: The `desc-` entity was not indexed because the `bids_dataframe.csv` file contained only the names of those files in which the filename was NOT changed (i.e. the entity `desc-`, wasn't added). Ivadomed training progresses without any error by conveniently discarding the `desc-` entity files.

**Conclusion (Preliminary)**: This approach of only adding the `desc-` entity (and the `DatasetType` key) without modifying the folder structure (to appropriately reflect the 2nd example mentioned in the discussion above) does not help because the modified files are unable to be indexed. 

**TODO**: Think of a different approach, this time possibly with a re-organization of the folder structure. I believe this can help because currently ivadomed is not able to locate these derivative files (i.e. the ones with the `desc-` entity) so placing these derivatives in the correct folder structure should solve this indexing issue.

**Comment by mariehbourget** [comment]
> **TODO**: Think of a different approach, this time possibly with a re-organization of the folder structure. I believe this can help because currently ivadomed is not able to locate these derivative files (i.e. the ones with the `desc-` entity) so placing these derivatives in the correct folder structure should solve this indexing issue.

Thank you @naga-karthik for the tests and detailed report.

I agree that re-organizing the structure would help indexing the files because `pybids` index _everything_ that is in the derivatives folder regardless of the naming. Unfortunately, `pybids` doesn't parsed the filenames of derivatives and the columns of the `bids_dataframe` associated with these files would be empty (as far as I  know).

Moreover, `ivadomed` needs "primary" `sub-XX` folders at the root of the dataset in its current implementation. Pointing to `awesome-data/derivatives/preprocessing-pipeline` as `path_data` would alleviate that but we would end up with the same indexing issue as before.

My guess is either:
1. We don't understand how to properly create a derived stand-alone dataset
2. `pybids` does not index those files in any cases when the validator is on.

From [here](https://github.com/bids-standard/pybids/blob/652074841cc922d6042fc500b14aabe438d42404/bids/layout/layout.py#L75) in `pybids`, it uses only the `bids.json` file by default and not the `derivatives.json` file. So the suggestion from @hermancollin to add `desc-` to our custom `config_bids.json` file was right. Another option would be to pass both the `bids.json` and `derivatives.json` as a list in the `bids_config` parameter in the `ivadomed` config file. Unfortunately, I think these approaches would require we turn off the validator as well.

**Comment by mariehbourget** [comment]
Update:
A while ago, we were wondering if having a `derivatives` folder inside a `derivatives` folder (Example case 1 in this [comment](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-879952192)) was allowed by BIDS.
The answer is YES, as described in the third example (`Example of a pipeline with nested derivative directories:`) in the spec [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets).

**Comment by mariehbourget** [comment]
I'm starting to work on this issue again to better understand the file indexing of "derivatives folder" and of "stand-alone derived datasets" by pybids.
I will document my findings here as I go but it may be a bit difficult to follow/see the big picture, so I will make a recapitulative comment when my investigations are finished.

**Comment by mariehbourget** [comment]
#### Clarify the indexation of "derived dataset" with `desc-` entity

Previous investigations (https://github.com/ivadomed/ivadomed/issues/860#issuecomment-880229411, https://github.com/ivadomed/ivadomed/issues/860#issuecomment-885987369) have shown that pybids was not indexing files with the `desc-` entity and/or adding its column/value to the dataframe, even when the `DatasetType` was set to `derivative` in `dataset_description.json`.

I suspect that there was an update in pybids since, because with the latest version `pybids==0.15.4`, it works.
I tested 2 cases:

1. A derived dataset where all files have the `desc-` entity in the filename.
2. A dataset where SOME files have the `desc-` entity and other files have a normal `raw data` naming (this is not supposed to happen in BIDS and is not compliant but was helpful for investigations).

In both cases, all files in **subject folders** (at the root of the dataset) were indexed properly by pybids provided that the 2 following fields were present in the root `dataset_description.json` file:
* `"DatasetType": "derivative",`
* `"GeneratedBy": [{"Name": "Test pipeline"}],`

This is true for both `anat` and `micr` datatypes.
Furthermore, a `desc` column appears in the dataframe with the correct values.
JSON metadata fields are included in the dataframe as expected (same as with "raw data").

Note: `micr` files are force indexed on master branch at the moment of testing (commit 487df8a).
So I tested also without the force indexing (not required since pybids update to include microscopy) on branch `mhb/update-force-index-micr-ct` of PR #1203 (commit b3c7a8ca).

**Comment by mariehbourget** [comment]
### Can we have a derivatives folder in a derived dataset?

**1. With the ivadomed naming convention for derivatives** (ex: `_seg-manual`) --> YES, the indexing and pairing of files is working as usual.

For example with:
    * filename at the sub-level of the derived dataset: `sub-XX_desc-YY_T2w.nii.gz`
    * filename at the derivatives level: `sub-XX_desc-YY_T2w_seg-manual.nii.gz`

**2. With the BIDS derivatives naming convention** --> NO, we will encounter some incompatibility if we already use the `desc-` entity in the sub-level filenames because the "mask" derivatives :
    * MAY contain the `desc-` entity as well but we cannot have 2 `desc-` entities in the same filename.
    * the original source entities MUST be before the `label-` entity but the `desc-` entity MUST come after the `label-` entity (ref: `<source_entities>[_space-<space>][_res-<label>][_den-<label>][_label-<label>][_desc-<label>]_mask.nii.gz`, [here](https://bids-specification.readthedocs.io/en/stable/05-derivatives/03-imaging.html#masks))

For example with:
    * filename at the sub-level of the derived dataset: `sub-XX_desc-YY_T2w.nii.gz`
    * filename at the derivatives level: `sub-XX_desc-YY_label-seg_mask.nii.gz` or `sub-XX_desc-YY_label-seg_desc-T2w_mask.nii.gz` are not BIDS compliant.

We may need to reach out to the BIDS community to better understand how to deal with "derivatives chains" as per the example in point 2.

**Comment by mariehbourget** [comment]
###  Can we have a derivatives folder in a derivatives folder?

A derivatives folder in a derivatives folder is compliant with BIDS as per: https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#storage-of-derived-datasets

Example
```
awesome-dataset/derivatives/awesome-dataset-processed/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz
awesome-dataset/derivatives/awesome-dataset-processed/derivatives/labels/sub-bob/anat/sub-bob_desc-XX_T1w_seg-manual.nii.gz
```
(I'm ignoring the file naming issues from the [previous comment](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-1309078451) with the BIDS derivatives convention and using the ivadomed convention for simplicity)

In ivadomed, it causes 2 minor issues:
1. The bids_dataframe filtering is based on the word "derivatives" being in the path or not. Having 2 "derivatives" string in the path is breaking [here](https://github.com/ivadomed/ivadomed/blob/ba7a30d32c991e83e0d46cc5e848e9033a769975/ivadomed/loader/bids_dataframe.py#L126-L134), so we would need to update the filtering to avoid this.
2. In order to ivadomed to work, we would need to set the `path_data` to the `awesome-dataset-processed`  folder (in the first derivatives folder), and we would need to copy the `participants.tsv` and `samples.tsv` there. I don't know if that would be considered BIDS compliant as I did not find clear restrictions to do so in the [Common principles](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html) or [Modality agnostic files](https://bids-specification.readthedocs.io/en/stable/03-modality-agnostic-files.html) sections.

Again, we may need to reach out to the BIDS community to know if modality agnostic files are allowed in derivatives pipeline folders. For reference, it doesn't raise errors in the validator 1.9.9.

**Comment by mariehbourget** [comment]
###  Clarify the indexation of "derivatives folder" (are files recognized and indexed? (YES) are the filename entities of derivatives included in the dataframe? (NO), how to make this happen?).

### Are JSON metadata in derivatives folder indexed in the dataframe?

* Currently, we use the `pybids.BIDSLayout` to index the root folder of the dataset with `derivatives=True`. In that case, derivatives files are indexed in the dataframe but only their path. The filename entities are not parsed and derivative JSON metadata (if present) is not indexed.
* Example with 2 T2w subjects and seg-manual label:

|path|data type|extension|subject|suffix|some TSV metadata|raw data JSON metadata|derivative JSON metadata|
|-|-|-|-|-|-|-|-|
|dataset_path/sub-cardiff03_T2w.nii.gz|anat|.nii.gz|cardiff03|T2w|AA|XX||
|dataset_path/sub-hamburg01_T2w.nii.gz|anat|.nii.gz|hamburg01|T2w|AA|XX||
|derivatives_path/sub-cardiff03_T2w_seg-manual.nii.gz||||||||
|derivatives_path/sub-hamburg01_T2w_seg-manual.nii.gz||||||||

* A possible workaround for this would be to use the `pybids.BIDSLayout` to index the root folder and the derivative folder separately with `derivatives=False` then combined the 2 dataframes together.
* In addition to parsing ALL entities and JSON in the dataframe, it could also simplify the code for the `bids_dataframe` filtering [here](https://github.com/ivadomed/ivadomed/blob/5b2e313154e94bbcc350ab56be7adf10f3d089c4/ivadomed/loader/bids_dataframe.py#L117-L131) and we could also assigned TSV metadata to derivatives.
* Example with 2 T2w subjects and seg-manual label:

|path|data type|extension|subject|suffix|some TSV metadata|raw data JSON metadata|derivative JSON metadata|
|-|-|-|-|-|-|-|-|
|dataset_path/sub-cardiff03_T2w.nii.gz|anat|.nii.gz|cardiff03|T2w|AA|XX||
|dataset_path/sub-hamburg01_T2w.nii.gz|anat|.nii.gz|hamburg01|T2w|AA|XX||
|derivatives_path/sub-cardiff03_T2w_seg-manual.nii.gz|anat|.nii.gz|cardiff03|manual|AA||YY|
|derivatives_path/sub-hamburg01_T2w_seg-manual.nii.gz|anat|.nii.gz|hamburg01|manual|AA||YY|

* Note: in pybids the `suffix` column for derivatives is based on the last underscore (`_`) in the filenames, so with the ivadomed naming convention for labels, we end up with `manual` in that column.

#### How is this related to the current issue?

* If we choose to use the concept of "derived dataset" for our file naming with for example the `desc-` entity, the same cnocept applied as in this [previous comment](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-1293951092) which is: both the derived `dataset root folder` and the `derivatives/labels folder `need to have a `dataset_description.json` file with the following fields in them:
    * `"DatasetType": "derivative",`
    * `"GeneratedBy": [{"Name": "Test pipeline"}],`
* In addition, our current implementation ask for a `dataset_description.json` file in the `derivatives` folder  OR in `derivatives/labels` [here](https://github.com/ivadomed/ivadomed/blob/5b2e313154e94bbcc350ab56be7adf10f3d089c4/ivadomed/loader/bids_dataframe.py#L323). If we want to index both the root folder and the labels folder separately, the `dataset_description.json` MUST be in `derivatives/labels`.

**Comment by mariehbourget** [comment]
### Are there additional JSON files/fields that are REQUIRED with the usage of desc-?

The usage of `desc-`, as intended in the initial post of this issue is the case of:
1.  [Preprocess or cleaned data](https://bids-specification.readthedocs.io/en/stable/05-derivatives/02-common-data-types.html#preprocessed-or-cleaned-data) and spedified that `"All REQUIRED metadata fields coming from a derivative file’s source file(s) MUST be propagated to the JSON description of the derivative unless the processing makes them invalid"`.


OR in the case of:

2. [Preprocessed, coregistered and/or resampled volumes](https://bids-specification.readthedocs.io/en/stable/05-derivatives/03-imaging.html#preprocessed-coregistered-andor-resampled-volumes), where the only field required is [SkullStripped](https://bids-specification.readthedocs.io/en/stable/glossary.html#objects.metadata.SkullStripped), true of false.

**Comment by mariehbourget** [comment]
### Summary

#### Context
This issue was initially open to explore the possibility of using the BIDS concept of "derived dataset" for our "preprocessed" datasets. We currently use the BIDS "raw data" naming scheme instead of the "derivatives" naming scheme for such datasets.
Please see [initial comment](https://github.com/ivadomed/ivadomed/issues/860#issue-943568439) for complete context.

Aside from the original questions, this investigation led to various questions regarding the indexing of derivatives files with `pybids`. More than perfect BIDS compliance, the `pybids` behavior is ultimately the important part for us as we are using it in ivadomed to index our datasets.

In parallel, we started looking into using the "derivatives" naming scheme for our "labels" instead of our own ivadomed convention in #1202. While this is not exactly the same issue, a lot of the investigations done here could be useful for the implementation in #1202.

As a reference, the newest batch of investigation starts [here](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-1293915167), where 6 topics related to derived dataset are covered.

#### Recommendations

* I would not recommend using the "derivatives" naming scheme for our preprocessed dataset until we hear back from the BIDS community regarding proper naming (re: https://github.com/ivadomed/ivadomed/issues/860#issuecomment-1309078451)
* Similarly, the future project of using the "derivatives" naming scheme for our label in #1202 may influence decisions here and should be treated in parallel.
* As for indexing derivatives in general, [this part](https://github.com/ivadomed/ivadomed/issues/860#issuecomment-1332317151) of the investigation could be implemented separately and improve the current bids_dataframe indexation.
